src/main/java/org/datadog/jenkins/plugins/datadog/util/CircuitBreaker.java

```diff
@@ -0,0 +1,83 @@
+package org.datadog.jenkins.plugins.datadog.util;
+
+import java.util.function.Consumer;
+import javax.annotation.concurrent.GuardedBy;
+
+public class CircuitBreaker<T> {
+
+    private static final int DEFAULT_MIN_HEALTH_CHECK_DELAY_MILLIS = 1000;
+    private static final int DEFAULT_MAX_HEALTH_CHECK_DELAY_MILLIS = 60000;
+    private static final double DEFAULT_DELAY_FACTOR = 2.0;
+
+    private final Consumer<T> action;
+    private final Consumer<T> fallback;
+    private final Consumer<Exception> errorHandler;
+    private final long minHealthCheckDelayMillis;
+    private final long maxHealthCheckDelayMillis;
+    private final double delayFactor;
+
+    @GuardedBy("this")
+    private boolean healthy;
+    @GuardedBy("this")
+    private long healthCheckDelayMillis;
+    @GuardedBy("this")
+    private long healthCheckAt;
+
+    public CircuitBreaker(Consumer<T> action, Consumer<T> fallback, Consumer<Exception> errorHandler) {
+        this(action, fallback, errorHandler, DEFAULT_MIN_HEALTH_CHECK_DELAY_MILLIS, DEFAULT_MAX_HEALTH_CHECK_DELAY_MILLIS, DEFAULT_DELAY_FACTOR);
+    }
+
+    public CircuitBreaker(Consumer<T> action,
+                          Consumer<T> fallback,
+                          Consumer<Exception> errorHandler,
+                          long minHealthCheckDelayMillis,
+                          long maxHealthCheckDelayMillis,
+                          double delayFactor) {
+        this.action = action;
+        this.fallback = fallback;
+        this.errorHandler = errorHandler;
+        this.minHealthCheckDelayMillis = minHealthCheckDelayMillis;
+        this.maxHealthCheckDelayMillis = maxHealthCheckDelayMillis;
+        this.delayFactor = delayFactor;
+        synchronized (this) {
+            this.healthy = true;
+            this.healthCheckDelayMillis = minHealthCheckDelayMillis;
+        }
+    }
+
+    public synchronized void accept(T t) {
```

Comment:
Do we expect a bottle neck here? Have we checked this with a high concurrency pipelines? I'm concerned about this would impact to the customer pipelines duration.

Reply:
Currently it's only used by the trace writer, and accept is only called from the polling thread, so this lock will never experience any contention.
The object is created in one thread and that used in the other. Strictly speaking we could do without synchronisation here (because of the HP-edge between calling thread start, and the first statement in its run method), but it felt too fragile to rely on these implementation details so I added explicit synchronisation.